examples/apps/table-detail-tea/src/main.rs

```diff
@@ -0,0 +1,239 @@
+use std::io::stdout;
+/// A Ratatui example that demonstrates the Elm architecture with a basic list - detail
+/// application.
+///
+/// This example runs with the Ratatui library code in the branch that you are currently
+/// reading. See the [`latest`] branch for the code which works with the most recent Ratatui
+/// release.
+///
+/// [`latest`]: https://github.com/ratatui/ratatui/tree/latest
+use std::time::Duration;
+
+use crossterm::event::{
+    self, DisableMouseCapture, EnableMouseCapture, Event, KeyCode, MouseEventKind,
+};
+use crossterm::ExecutableCommand;
+use ratatui::layout::{Constraint, Layout};
+use ratatui::style::{Modifier, Style};
+use ratatui::text::Line;
+use ratatui::widgets::{Block, Borders, Padding, Paragraph, Row, Table, TableState};
+use ratatui::Frame;
+
+static INFO_TEXT: &str = " (Esc/q) quit | (↑) move up | (↓) move down ";
+
+/// Application data model and state
+#[derive(Debug, Default)]
+struct Model {
+    table_items: Vec<Data>,
+    table_state: TableState,
+    running_state: RunningState,
+}
+
+#[derive(Debug, Default, PartialEq, Eq)]
+enum RunningState {
+    #[default]
+    Running,
+    Done,
+}
+
+/// Data model
+#[derive(Debug, Default)]
+struct Data {
+    name: String,
+    address: String,
+    email: String,
+}
+
+#[derive(PartialEq, Copy, Clone)]
+enum Message {
+    SelectNext,
+    SelectPrevious,
+    Quit,
+}
+
+fn main() -> color_eyre::Result<()> {
+    tui::install_panic_hook();
+    let mut terminal = tui::init_terminal()?;
+    let mut model = Model {
+        table_items: generate_some_fans(),
+        ..Default::default()
+    };
+    stdout().execute(EnableMouseCapture)?;
+    while model.running_state != RunningState::Done {
+        // Render the current view
+        terminal.draw(|f| view(&mut model, f))?;
+
+        // Handle events and map to a Message
+        let mut current_msg = handle_event(&model)?;
+
+        // Process updates as long as they return a non-None message
+        while current_msg.is_some() {
+            current_msg = update(&mut model, current_msg.unwrap());
+        }
+    }
```

Comment:
Failing in the main part of the app (and propagating the error using `?`) should not cause the restore terminal code to run. The easiest way to avoid this problem is to extract this loop to a method (e.g. name it run) and store the result, but only evaluate it after restoring the terminal). See the rest of the examples for this pattern.

Reply:
Please adjust as you see fit. Copied it from: https://ratatui.rs/concepts/application-patterns/the-elm-architecture/